pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BucketDelayedDeliveryTracker.java

```diff
@@ -243,6 +263,53 @@ public synchronized boolean addMessage(long ledgerId, long entryId, long deliver
         return true;
     }
 
+    private synchronized CompletableFuture<Void> asyncMergeBucketSnapshot() {
+        List<ImmutableBucket> values = immutableBuckets.asMapOfRanges().values().stream().toList();
+        long minNumberMessages = Long.MAX_VALUE;
+        int minIndex = -1;
+        for (int i = 0; i + 1 < values.size(); i++) {
+            ImmutableBucket bucketL = values.get(i);
+            ImmutableBucket bucketR = values.get(i + 1);
+            long numberMessages = bucketL.numberBucketDelayedMessages + bucketR.numberBucketDelayedMessages;
+            if (numberMessages < minNumberMessages) {
+                minNumberMessages = (int) numberMessages;
+                minIndex = i;
+            }
+        }
+        return asyncMergeBucketSnapshot(values.get(minIndex), values.get(minIndex + 1));
+    }
+
+    private synchronized CompletableFuture<Void> asyncMergeBucketSnapshot(ImmutableBucket bucketA,
+                                                                          ImmutableBucket bucketB) {
+        immutableBuckets.remove(Range.closed(bucketA.startLedgerId, bucketA.endLedgerId));
+        immutableBuckets.remove(Range.closed(bucketB.startLedgerId, bucketB.endLedgerId));
+
+        CompletableFuture<Long> snapshotCreateFutureA =
+                bucketA.getSnapshotCreateFuture().orElse(CompletableFuture.completedFuture(null));
+        CompletableFuture<Long> snapshotCreateFutureB =
+                bucketB.getSnapshotCreateFuture().orElse(CompletableFuture.completedFuture(null));
+
+        return CompletableFuture.allOf(snapshotCreateFutureA, snapshotCreateFutureB).thenCompose(__ -> {
+            CompletableFuture<List<DelayedMessageIndexBucketSnapshotFormat.SnapshotSegment>> futureA =
+                    bucketA.getRemainSnapshotSegment();
+            CompletableFuture<List<DelayedMessageIndexBucketSnapshotFormat.SnapshotSegment>> futureB =
+                    bucketB.getRemainSnapshotSegment();
+            return futureA.thenCombine(futureB, CombinedSegmentDelayedIndexQueue::wrap)
+                    .thenCompose(combinedDelayedIndexQueue -> {
+                        CompletableFuture<Void> removeAFuture = bucketA.asyncDeleteBucketSnapshot();
+                        CompletableFuture<Void> removeBFuture = bucketB.asyncDeleteBucketSnapshot();
+
+                        return CompletableFuture.allOf(removeAFuture, removeBFuture).thenRun(() -> {
```

Comment:
We do save the newly merged bucket after finished delete the original buckets, it may be caused bucket loss. I think losing buckets is tolerable, right?

Reply:
Maybe we can persist the newly merged bucket, then delete the original buckets.

Reply:
If the bucket index is lost, the tracker needs to rebuild the index, but if there are two intersecting buckets, it will cause data confusion.

Reply:
1. Delete the original bucket index first, then save the newly merged index, this may lose the bucket index, but the tracker will rebuild the index.
2. Save the newly merged index first, then delete the original bucket index, this may cause intersecting buckets, and we need to find a way to clean up the original buckets.

I think there are two ways, the first way is simple and clear.

Reply:
Question: 
When is the index rebuilt?
Should we unload the topic to rebuild the index if this operation fails?

Reply:
> When is the index rebuilt?

If index is lost, the tracker will rebuild index when reread by the managed cursor.

> Should we unload the topic to rebuild the index if this operation fails?

Yes, we can unload the topic to rebuild the index

Reply:
Can we mark the old segment as deleting or merging? So that we can continue the merge operation after the broker crashes.

Reply:
> Can we mark the old segment as deleting or merging? So that we can continue the merge operation after the broker crashes.

This will introduce more state in metadata, we can improve it when we really need it

Reply:
I use a way to clean up the old overlap buckets when recovering buckets, so that we can persist the newly merged bucket first, then delete the original bucket.